djtriggers/models.py

```diff
@@ -75,35 +82,111 @@ def __init__(self, *args, **kwargs):
 
-    def process(self, force=False, logger=None, dictionary={}):
-        if logger:
-            self.logger = get_logger(logger)
-        now = timezone.now()
-        if not force and self.date_processed is not None:
-            raise AlreadyProcessedError()
-        if not force and self.process_after and self.process_after >= now:
-            raise ProcessLaterError(self.process_after)
-
+    def process(self, force=False, logger=None, dictionary=None, use_statsd=False):
+        """
+        Executes the Trigger
+        :param bool force: force the execution
+        :param string logger: slug of preferred logger
+        :param dict dictionary: dictionary needed by trigger to execute
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        if dictionary is None:
+            dictionary = {}
+
+        # The task gets locked because multiple tasks in the queue can process the same trigger.
+        # The lock assures no two tasks can process a trigger simultaneously.
+        # The check for date_processed assures a trigger is not executed multiple times.
         try:
-            self.logger.log_result(self, self._process(dictionary))
-        except ProcessLaterError as e:
-            self.process_after = e.process_after
-            self.save()
-            raise
-        if self.date_processed is None:
-            self.date_processed = now
-        self.save()
+            with NonBlockingLock.objects.acquire_lock(self):
+                if logger:
+                    self.logger = get_logger(logger)
+                now = timezone.now()
+                if not force and self.date_processed is not None:
+                    raise AlreadyProcessedError()
+                if not force and self.process_after and self.process_after >= now:
+                    raise ProcessLaterError(self.process_after)
+
+                try:
+                    # execute trigger
+                    self.logger.log_result(self, self._process(dictionary))
+                    self._handle_execution_success(use_statsd)
+                except ProcessLaterError as e:
+                    self.process_after = e.process_after
+                    self.save()
+                    raise
+                except Exception as e:
+                    self._handle_execution_failure(e, use_statsd)
+                    raise
+        except AlreadyLocked:
+            pass
 
     def _process(self, dictionary):
         raise NotImplementedError()
 
+    def _handle_execution_failure(self, exception, use_statsd=False):
+        """
+        Handle execution failure of the trigger
+        :param Exception exception: the exception raised during failure
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        self.number_of_tries += 1
+        if self.number_of_tries >= getattr(settings, 'DJTRIGGERS_TRIES_BEFORE_WARNING', 3):
```

Comment:
there is an off by one here since we increment before checking and include equals in the check. if `tries_before_warning = 3` it will only be tried twice.